chanzuckerberg/corpora-data-portal
# use react-query

This PR adds [`react-query`](https://github.com/tannerlinsley/react-query) to be our server state management library, instead of using Redux Toolkit to do both client state and server state management. The benefits of this arrangement include:

1. Instead of replicating server schemas and entity tables, and trying to normalize entities and manage entity updates in the FE, we rely on straight forward query cache invalidations to signal `react-query` to atomically refetch server data when needed, since BE is and should be the single source of truth for server state, instead of having FE maintaining and managing a snapshot of server state, which is troublesome and inefficient

1. React Query comes with really awesome cache management, garbage collection, and convenient auto refetch [default configuration](https://react-query.tanstack.com/docs/guides/important-defaults). This means unused server state will be cleared from the browser memory after a certain time and refetch will happen in the background for us, such as when the window refocuses, internet reconnects, etc..

1. Most importantly, React Query provides a really nice developer experience (also has [React Query Devtools](https://github.com/tannerlinsley/react-query-devtools)) and reduces even more code than what Redux Toolkit has already achieved compared to vanilla Redux

1. Note that React Query is **NOT** a replacement for client state management, which can be still done via Redux Toolkit and/or plain React Context/Hooks. At this stage of our app, we don't need to introduce Redux Toolkit yet

Supplemental resources:
1. [React Query release tech talk](https://www.youtube.com/watch?v=seU46c6Jz7E)
1. [React Query docs](https://react-query.tanstack.com/docs/overview)
1. [My notes on React Query](https://docs.google.com/document/d/1G-pxb2NoUImo-aa9fWmz91K8jaeQTHNT_0CDXyqSeeE/edit)

Code change:
1. Refactor current `ProjectList` to use React Query
1. Rename `ProjectList` to `Collections`
1. Extract dataset related API calls from `Homepage` into `Collections`
1. Add [React Query Devtools](https://github.com/tannerlinsley/react-query-devtools)
1. Add `frontend/src/common/queries/README.md` to explain file structure, minimal best practices, and some react query resources

QA notes:
1. Please make sure the datasets and still rendered correctly
1. NOTE: you will notice that `react-query` by default will refetch the data in the background when the window is refocused. Pretty cool!
![gif](https://user-images.githubusercontent.com/6309723/97214392-4a9d2f00-1780-11eb-977e-9bb3cdc8389d.gif)
